### PR TITLE
added some split configurability to header validation plugin

### PR DESCRIPTION
**Pull request by iamnotbrandonchan**
This PR addresses SID-96.
Problem: There is hardcoded "," as a delimiter for tokenizing headers in the validate header. For example, if a header was: ldap-groups: "development,sre", we can split into "development", "sre". However, suppose the header was ldap-groups: "development-sre". We have no configuration for tokenizing.

Solution: Add a tokenizer config for specific headers -> more configuration yaml LOLOL😆

### CODE REVIEW

File: pkg/components/validate_headers.go [hunk 1]
```diff
@@ -24,16 +25,21 @@ func contains(s []string, target string) bool {
...
 
-func incomingMatchesAllowed(allowed map[string][]string, incoming map[string][]string) error {
+func incomingMatchesAllowed(allowed map[string][]string, incoming map[string][]string, split map[string]string) error {
 	checkedHeaderAndValues := make(map[string][]string)
-	for allowedKey, allowedValues := range allowed {
+	for allowedHeader, allowedValues := range allowed {
 		// check if incoming header values have a configured allowed header present and search through them if so
-		if matchedIncomingHeaderValues, present := incoming[http.CanonicalHeaderKey(allowedKey)]; present {
+		if matchedIncomingHeaderValues, present := incoming[http.CanonicalHeaderKey(allowedHeader)]; present {
+			delimiter := " "
```
Comment (iamnotbrandonchan) [inline pkg/components/validate_headers.go]: Question: I'm not happy about this part, specifically setting a default of " ". What I really want to do is set this default to empty string, "". However, if i do this, then within the contains function, we split on "", which results in ->
"split" -> "s","p","l","i","t". This is the best i could come up with. Does anyone have any suggestions? :)

Comment (faldridge) [reply pkg/components/validate_headers.go]: Yes, I would assume if no delimiter is explicitly set, then the entire value should be used, without splitting.

Suggestion: update the code such that the allowedValues are all checked against the entire string when no explicit delimiter is specified. 

Of course, this means the only combination that makes sense or that would ever work when no delimiter is given is a single allowed value, but that seems acceptable to me. 🤷‍♂

Comment (mikerott) [reply pkg/components/validate_headers.go]: Agree with @faldridge.

File: pkg/components/validate_headers.go [hunk 2]
```diff
@@ -10,30 +10,41 @@ import (
...
-		if matchedIncomingHeaderValues, present := incoming[http.CanonicalHeaderKey(allowedKey)]; present {
+		if matchedIncomingHeaderValues, present := incoming[http.CanonicalHeaderKey(allowedHeader)]; present {
+			delimiter := ""
+			if splitValue, ok := split[allowedHeader]; ok {
+				delimiter = splitValue
+			}
+
 			// iterate through configured allowed values for a header
-			for _, item := range allowedValues {
+			for _, allowedValue := range allowedValues {
```
Comment (faldridge) [inline pkg/components/validate_headers.go]: 👍

File: pkg/components/validate_headers.go [hunk 3]
```diff
@@ -53,6 +64,7 @@ func (r *validateHeaderTransport) RoundTrip(req *http.Request) (*http.Response,
 // ValidateHeaderConfig is used to validate a map of headers and their allowed values against an incoming requests headers
 type ValidateHeaderConfig struct {
 	Allowed map[string][]string `description:"Map of headers to validate and the allowed values for those headers."`
+	Split   map[string]string   `description:"Map of headers to split on the associated value"`
```
Comment (faldridge) [inline pkg/components/validate_headers.go]: Suggestion: tweak the wording on the `Split` description to `Map of headers to the delimiter character to use for headers that contain multiple values.`

**Comment by mikerott** [comment]
I changed this PR because the settings lib does not (yet) support `map[string]string` though it would be somewhat easy to put it in.  Anyway, now the PR represents a `split: ","` config param one can apply to the `validateheaders` plugin.  If `split` is omitted or empty string, the code defaults back to `,`.  The code disallows strings 2 or more chars long.  See unit tests.  I also updated the README docs and integration tests.

File: README.md [hunk 4]
```diff
@@ -420,7 +429,8 @@ func main() {
 	plugins := []transportd.NewComponent{
 		components.Metrics,
 		components.AccessLog,
-		components.ASAPValidate,
+    components.ASAPValidate,
```
Comment (iamnotbrandonchan) [inline README.md]: Nit: can you put make these 2 lines consistent with the others in terms of indentation

Comment (ensary) [inline README.md]: Nit: tabs vs speaces (the file is formatted with tabs, the change has spaces)

File: pkg/components/validate_headers.go [hunk 5]
```diff
@@ -75,10 +83,19 @@ func (*ValidateHeaderConfigComponent) Settings() *ValidateHeaderConfig {
 
 // New generates the middleware.
 func (*ValidateHeaderConfigComponent) New(_ context.Context, conf *ValidateHeaderConfig) (func(tripper http.RoundTripper) http.RoundTripper, error) {
+	if strings.TrimSpace(conf.Split) == "" { // revert to default
+		conf.Split = ","
```
Comment (iamnotbrandonchan) [inline pkg/components/validate_headers.go]: Question: Do we want the default to still split? Shouldn't the default not split? This one is debatable...

Comment (ensary) [reply pkg/components/validate_headers.go]: Default to no split will make this a breaking change.

**Comment by gcase555** [comment]
@mikerott I believe `map[string]string` is supported?

File: pkg/components/validate_headers.go (hunk 5 above)
Comment (gcase555) [reply pkg/components/validate_headers.go]: @ensary @iamnotbrandonchan This being a breaking change could be fine if we create a new version, I think defaulting to split could lead to some scenarios where it is undesirable

Comment (ensary) [reply pkg/components/validate_headers.go]: 👍 , especially if we will update the existing deployment in a timely fashion.

File: pkg/components/validate_headers.go (hunk 2 above)
Comment (gcase555) [reply pkg/components/validate_headers.go]: agreed! much improved!

**Comment by gcase555** [comment]
Support for `map[string]string` was added to settings here in this commit https://github.com/asecurityteam/settings/commit/b8b4e8355e0d572a3dc40061c089d7e8d22edc7c @iamnotbrandonchan @mikerott

File: pkg/components/validate_headers.go (hunk 5 above)
Comment (gcase555) [reply pkg/components/validate_headers.go]: it's possible we could do the breaking change in a separate PR though

**Comment by gcase555** [comment]
FYI reviewers myself and @iamnotbrandonchan will work on this on friday to try to get the configurable delimiter character to be per character and address some of the comments here

File: README.md [hunk 6]
```diff
@@ -198,6 +199,14 @@ x-transportd:
...
     # ([]string) Acceptable issuer strings.
     allowedissuers:
+  validateheaders:
+    # (map[string][]string) allowed list of headers whose values to check
+    allowed:
+      accept:
+        - "text/json"
+        - "text/html"
+    # (string) the delimiter to use for splitting header values when they come in single line
+    split: ","
```
Comment (gcase555) [inline README.md]: task: update this example in the readme, either remove it or use a simple example like what's in the `/docs` directory

Comment (iamnotbrandonchan) [reply README.md]: good call, will fix it now

File: docs/components.md [hunk 7]
```diff
@@ -21,8 +21,14 @@ validateheaders:
       - "pm"
 ```
 
+`validate_headers` also supports splitting on a configurable delimiter for each `allowed` header. To accomplish this, we set a `split` configuration using the headers name and the character(s) we want to `split` on.
```
Comment (ensary) [inline docs/components.md]: Do we allow multi-char?

Comment (iamnotbrandonchan) [reply docs/components.md]: yep! we allow it. I can't think of a case where we not want to allow multi-char. YAGNI. And if this case arises in the future, we can iterate on this

File: README.md [hunk 8]
```diff
@@ -198,6 +199,15 @@ x-transportd:
...
     # ([]string) Acceptable issuer strings.
     allowedissuers:
+  validateheaders:
+    # (map[string][]string) allowed list of headers whose values to check
+    allowed:
+      accept:
+        - "text/json"
+        - "text/html"
+    # (map[string] string) the delimiters to use for splitting header-specific values when they come in single line
+    split:
```
Comment (gcase555) [inline README.md]: 👍

File: docs/components.md (hunk 7 above)
Comment (gcase555) [reply docs/components.md]: we do support multi-char splits at this time and wrote some tests to verify this behavior works

File: pkg/components/validate_headers_test.go [hunk 9]
```diff
@@ -10,82 +10,95 @@ import (
 
 func Test_incomingMatchesAllowed(t *testing.T) {
 	defaultAllowedHeaderAndValues := map[string][]string{"Ldap-Groups": {"sre", "devs"}}
+	defaultDelimiterValues := map[string]string{"Ldap-Groups": ","}
 	tests := []struct {
 		name           string
 		allowedHeader  map[string][]string
 		incomingHeader map[string][]string
+		delimiter      map[string]string
```
[the remark this answers is not in the document]
Comment (gcase555) [reply pkg/components/validate_headers_test.go]: we have a multi-char test here https://github.com/asecurityteam/transportd/pull/34/files#diff-2d388de09264a551d71493584e8f0f48R214

**Pull request merged**